src/correction/cellbender_remove_background/helper.py

```diff
@@ -0,0 +1,137 @@
+                  'analyzed_barcodes_only=False')
+
+    # Construct the anndata object.
+    adata = anndata.AnnData(X=X,
+                            obs={'barcode': d.pop('barcodes').astype(str)},
+                            var={'gene_name': (d.pop('gene_names') if 'gene_names' in d.keys()
+                                               else d.pop('name')).astype(str)},
+                            dtype=X.dtype)
+    adata.obs.set_index('barcode', inplace=True)
+    adata.var.set_index('gene_name', inplace=True)
+
+    # For CellRanger v2 legacy format, "gene_ids" was called "genes"... rename this
+    if 'genes' in d.keys():
+        d['id'] = d.pop('genes')
+
+    # For purely aesthetic purposes, rename "id" to "gene_id"
+    if 'id' in d.keys():
+        d['gene_id'] = d.pop('id')
+
+    # If genomes are empty, try to guess them based on gene_id
+    if 'genome' in d.keys():
+        if np.array([s.decode() == '' for s in d['genome']]).all():
+            if '_' in d['gene_id'][0].decode():
+                print('Genome field blank, so attempting to guess genomes based on gene_id prefixes')
+                d['genome'] = np.array([s.decode().split('_')[0] for s in d['gene_id']], dtype=str)
+
+    # Add other information to the anndata object in the appropriate slot.
+    _fill_adata_slots_automatically(adata, d)
+
+    # Add a special additional field to .var if it exists.
+    if 'features_analyzed_inds' in adata.uns.keys():
+        adata.var['cellbender_analyzed'] = [True if (i in adata.uns['features_analyzed_inds'])
+                                            else False for i in range(adata.shape[1])]
+
+    if analyzed_barcodes_only:
+        for col in adata.obs.columns[adata.obs.columns.str.startswith('barcodes_analyzed')
+                                     | adata.obs.columns.str.startswith('barcode_indices')]:
+            try:
+                del adata.obs[col]
+            except Exception:
+                pass
+    else:
+        # Add a special additional field to .obs if all barcodes are included.
+        if 'barcodes_analyzed_inds' in adata.uns.keys():
+            adata.obs['cellbender_analyzed'] = [True if (i in adata.uns['barcodes_analyzed_inds'])
+                                                else False for i in range(adata.shape[0])]
+
+    return adata
+
+
+def dict_from_h5(file: str) -> Dict[str, np.ndarray]:
+    """Read in everything from an h5 file and put into a dictionary."""
+    d = {}
+    with tables.open_file(file) as f:
+        # read in everything
+        for array in f.walk_nodes("/", "Array"):
+            d[array.name] = array.read()
+    return d
+
+
+def _fill_adata_slots_automatically(adata, d):
+    """Add other information to the adata object in the appropriate slot."""
+
+    for key, value in d.items():
+        try:
+            if value is None:
+                continue
+            value = np.asarray(value)
+            if len(value.shape) == 0:
```

Comment:
This is already much better I think (can still be improved, but this is at least more readable)

```python
try:
    n_rows, n_cols, *multi_dim = value.shape
except ValueError:
    ﻿adata.uns[key] = value
else:
    if value.n_rows == adata.n_obs:
        if multi_dim or n_cols < 2:
            adata.obs[key] = value
         else:
             adata.obsm[key] = value
    elif n_rows == adata.n_vars:
        adata.var[key] = value.astype(str) if value.dtype.name.startswith('bytes') else value

```